aten/src/ATen/core/ATenDispatch.h

```diff
@@ -132,18 +146,104 @@ class CAFFE2_API ATenDispatch {
 
 CAFFE2_API ATenDispatch& globalATenDispatch();
 
+template<class Result, class... Args>
+Result callBoxedFallback(const char* schema, FallbackBoxedFunction* boxed_fallback_fn, Args&&... args,
+  // NB: enable_if must occur in function parameter, because MSVC
+  // doesn't like it when it's a template argument next to
+  // a parameter pack
+  typename c10::guts::enable_if_t<
+    !supports_boxed_fallback<Result, Args...>::value,
+    std::nullptr_t
+  > = nullptr) {
+  // This is dead because we test the SFINAE condition before calling
+  // boxed_fallback_fn.  A more functional way of writing this with
+  // optional<Result> return value works poorly when void is involved.
+  TORCH_INTERNAL_ASSERT(0);
+}
+
+template<
+  class Result, class... Args>
+Result callBoxedFallback(const char* schema, FallbackBoxedFunction* boxed_fallback_fn, Args&&... args,
+  typename c10::guts::enable_if_t<
+    supports_boxed_fallback<Result, Args...>::value,
+    std::nullptr_t
+  > = nullptr) {
+  torch::jit::Stack stack;
+  torch::jit::push(stack, std::forward<Args>(args)...);
+  boxed_fallback_fn(schema, &stack);
+  TORCH_INTERNAL_ASSERT(stack.size() == 1);
+  return torch::jit::pop(stack).to<Result>();
+}
+
+// NB: No universal forwarding
+template<class Result, class... Args>
+Result ATenOpTable::callUnboxed(Args... args) const {
+  using FuncType = Result(Args...);
+  // NB: No universal forwarding (takes const& only)
+  TensorTypeSet ts = detail::multi_dispatch_tensor_type_set(args...);
+  TensorTypeId tid = impl::dispatchTypeId(ts);
+
+  // You might think we can eliminate the second branch by maintaining a
+  // bitmask of registered operator keys, so we don't select dispatch ids
+  // which don't have implementations here.  But the net effect is that if you
+  // get a Variable CPUTensor, if there is no variable registration, you'll
+  // fall back to the CPU implementation.  Is this what you want?  Unlikely...
+
+  auto* unboxed_fn = reinterpret_cast<FuncType*>(function_table_[static_cast<int64_t>(tid)]);
+  if (C10_LIKELY(unboxed_fn != nullptr)) {
+    return (*unboxed_fn)(std::forward<Args>(args)...);
+  }
+
+  // The supports_boxed_fallback condition test, and the SFINAE on
+  // callBoxedFallback, do the same thing.  But we perform this (compile-time)
+  // test twice so that we can take advantage of the fact that return
+  // func_returns_void() is OK.  If we eliminated this condition in exchange
+  // for having callBoxedFallback return an optional, we can't conveniently
+  // handle the Result=void case anymore.
+  //
+  // (The SFINAE in callBoxedFallback, of course, is necessary to
+  // prevent us from attempting to typecheck code that won't typecheck.)
+  auto* boxed_fallback_fn = globalATenDispatch().getFallbackBoxedOp(tid);
+  if (C10_UNLIKELY(boxed_fallback_fn)) {
+    if (supports_boxed_fallback<Result, Args...>::value) {
+      return callBoxedFallback<Result, Args...>(schema_.c_str(), boxed_fallback_fn, std::forward<Args>(args)...);
```

Comment:
can we pass in a FunctionSchema rather than the schema string here?

did some light profiling and found dealing with this string on every op invocation is a huge overhead

Reply:
Not surprising :) Yes, FunctionSchema can probably be arranged.

Reply:
Recapping in person conversation: our new plan is to get FunctionSchema out of JIT data structures